api/lib/infrastructure/repositories/answer-repository.js

```diff
@@ -1,114 +1,134 @@
-const fp = require('lodash/fp');
+const _ = require('lodash');
+const jsYaml = require('js-yaml');
+const { knex } = require('../../../db/knex-database-connection');
+const { NotFoundError } = require('../../domain/errors');
 const Answer = require('../../domain/models/Answer');
 const answerStatusDatabaseAdapter = require('../adapters/answer-status-database-adapter');
-const BookshelfAnswer = require('../orm-models/Answer');
-const BookshelfKnowledgeElement = require('../orm-models/KnowledgeElement');
 
-const Bookshelf = require('../bookshelf');
-const { NotFoundError } = require('../../domain/errors');
-const jsYaml = require('js-yaml');
-const _ = require('lodash');
-
-function _adaptModelToDb(answer) {
+function _adaptAnswerToDb(answer) {
   return {
-    id: answer.id,
+    ..._.pick(answer, ['value', 'timeout', 'challengeId', 'assessmentId', 'timeSpent']),
     result: answerStatusDatabaseAdapter.toSQLString(answer.result),
     resultDetails: jsYaml.dump(answer.resultDetails),
-    value: answer.value,
-    timeout: answer.timeout,
-    challengeId: answer.challengeId,
-    assessmentId: answer.assessmentId,
-    timeSpent: answer.timeSpent,
   };
 }
 
-function _toDomain(bookshelfAnswer) {
-  if (bookshelfAnswer) {
-    return new Answer({
-      id: bookshelfAnswer.get('id'),
-      result: answerStatusDatabaseAdapter.fromSQLString(bookshelfAnswer.get('result')),
-      resultDetails: bookshelfAnswer.get('resultDetails'),
-      timeout: bookshelfAnswer.get('timeout'),
-      value: bookshelfAnswer.get('value'),
-      assessmentId: bookshelfAnswer.get('assessmentId'),
-      challengeId: bookshelfAnswer.get('challengeId'),
-      timeSpent: bookshelfAnswer.get('timeSpent'),
-    });
-  }
-  return null;
+function _adaptKnowledgeElementToDb(knowledgeElement) {
+  return _.pick(knowledgeElement, ['source', 'status', 'earnedPix',
+    'answerId', 'assessmentId', 'skillId', 'userId', 'competenceId']);
+}
+
+function _toDomain(answerDTO) {
```

Comment:
Attention, digression un peu pédante dans la forme, mais pas sur le fond (ne pas introduire de nouveaux concepts sauf si aucun existant ne peut être utilisé)

Je prèfère le terme Data structure pour plusieurs raisons
- il n'utilise pas le terme `object` qui est ambigü en JS: est-ce un objet au sens de la POO ? Ou est-ce au sens du "type" `Object` ?
- le terme DTO est utilisé par [Fowler](https://martinfowler.com/eaaCatalog/dataTransferObject.html) dans un contexte Java, et je ne sais pas si ce mot est passé en JS (je pense qu'on ne peut pas partir du principe que les développeurs de Pix ont tous fait du Java et pour lequel ce terme est plus clair que `data structure`)
- Robert Martin utilise le terme `data structure` qui est utilisé [depuis le début](https://en.wikipedia.org/wiki/Algorithms_%2B_Data_Structures_%3D_Programs) de l'informatique, avant l'OO

https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html

> Typically the data that crosses the boundaries is simple data structures. You can use basic structs or simple Data Transfer objects if you like. Or the data can simply be arguments in function calls. 


```suggestion
function _toDomain(answerDataStructure) {
```

Reply:
@octo-topi je t'invite à comparer https://en.wikipedia.org/wiki/Data_structure à https://en.wikipedia.org/wiki/Data_transfer_object.
Remarque que Martin parle de _simple_ data structures. Un DTO est exactement ça, un exemple de structure de données simple.
Enfin ce n'est pas la première fois qu'on parle de DTO dans Pix, si ça posait problème aux non-Javaistes on l'aurait remarqué plus tôt non ?

Reply:
> Enfin ce n'est pas la première fois qu'on parle de DTO dans Pix, 
Je n'en ai jamais vu, mais si ça parle à tout le monde, ça me va

Reply:
La seule mention aux DTO que je vois dans l'API est http://github.com/1024pix/pix/blob/dev/api/lib/infrastructure/repositories/jury-certification-summary-repository.js, il me manque un truc ou pas ?

Reply:
https://github.com/1024pix/pix/blob/0da62021dfcf86b72645256059dec5e8a0430aba/api/lib/application/certification-reports/certification-report-controller.js#L8-L9
https://github.com/1024pix/pix/blob/0da62021dfcf86b72645256059dec5e8a0430aba/api/lib/infrastructure/repositories/certification-course-repository.js#L16-L21
https://github.com/1024pix/pix/blob/0da62021dfcf86b72645256059dec5e8a0430aba/api/lib/domain/usecases/save-certification-issue-report.js#L6-L18
etc.

Reply:
J'avoue qu'en dehors du scope de certif, je n'ai pas vu ce terme employé :) 

Donc le mettre en dehors demande au moins un peu d'éclaircissement pour qu'il soit compris partout.

Reply:
Maintenant que tu le signales @MelanieMEB c'est vrai que presque toutes les occurrences que j'ai trouvées ont un rapport avec la certif. J'avoue que je préfère éviter de raisonner en termes de "scopes fonctionnels" quand il s'agit des standards de code 😉 

Cela étant, pour être précis le terme DTO est apparu dans le code Pix dans [ce commit](https://github.com/1024pix/pix/commit/58a22e84ade5b5f61febac10a8d6e97a173b00fd) qui n'a pas de rapport avec la certif 😉 — ironiquement, de façon incorrecte puisque la fonction en question nommée `_toStudentWithUserInfoDTO` renvoie une instance d'une classe chargée depuis `domain/models` !

La première utilisation "correcte" du terme est [ici](https://github.com/1024pix/pix/commit/fa8c59e8635803ec7fdf3530a1738da899277428) et c'est bien de la certif.

Noter aussi qu'on a quelques occurrences de `DataObject` depuis [avril 2018](https://github.com/1024pix/pix/commit/1d97ce300d43755e28bfc73a350c79e2eb4c3293), qui servent à distinguer les données "pures" du référentiel (ex. `ChallengeDataObject`), qui sont stockées notamment dans le cache (d'où la partie _transfer_), des objets avec du comportement qu'on construit en partant de là (ex. `Challenge`).

C'est plus à ce dernier exemple que je pensais quand je disais qu'on parlait déjà de DTO. Même si on n'avait pas mis le nom du [pattern](https://martinfowler.com/eaaCatalog/dataTransferObject.html) dans le code.

Reply:
Alors, le lien vers [les data structure](https://en.wikipedia.org/wiki/Data_structure) est éclairant

> There are numerous types of data structures, generally built upon simpler primitive data types:
- array
- list
- record
- union 
- tagged union
- object 

> An object is a data structure that contains data fields, like a record does, as well as various methods which operate on the data contents. In the context of object-oriented programming, records are known as plain old data structures to distinguish them from objects.

Dans le contexte JS, lorsque tu évoques les DTO, tu as en tête les "plain old data structures", à savoir des "Object" JS qui ne contiennent pas de méthodes, n'est-ce pas ?

J'ai aussi regardé le lien vers [DTO](https://en.wikipedia.org/wiki/Data_transfer_object). Il est très intéressant car il recentre le sujet sur la notion d'appel distant (client/serveur).
Autrement dit, une "plain old data structure" est une implémentation possible d'un pattern au niveau client/serveur; il pourrait aussi être utilisé dans un autre pattern à un autre niveau (ex: clean archi dans API REST).

>  Because the majority of the cost of each call is related to the round-trip time between the client and the server, one way of reducing the number of calls is to use an object (the DTO) that aggregates the data that would have been transferred by the several calls, but that is served by one call only.

Dans [un autre article de Fowler](https://martinfowler.com/bliki/LocalDTO.html) qui évoque ce point ambigü, le pattern me semble s'appliquer par exemple à l'interface d'une API REST, pas aux couches de l'archi hexa/clean archi qui la composent.

> DTOs are called Data Transfer Objects because their whole purpose is to shift data in expensive remote calls. They are part of implementing a coarse grained interface which a remote interface needs for performance. Not just do you not need them in a local context, they are actually harmful both because a coarse-grained API is more difficult to use and because you have to do all the work moving data from your domain or data source layer into the DTOs.

Bref, tout ça pour dire que peu importe les mots, du moment qu'on met globalement la même chose derrière..

Reply:
@EGI-OCTO toi qui as utilisé le terme DTO, quelle est ta définition du terme ?
https://github.com/1024pix/pix/commit/fa8c59e8635803ec7fdf3530a1738da899277428

Reply:
Data transfer object : une structure de données qui sert à faire la glue dans les endroits où on a besoin de serialiser/deserialiser (domain->bdd, bdd->domain, http->domain, domain->http, idem pour toutes les entrées/sorties et parfois aussi pour faire de la traduction/l'isolation entre domain/librairies et/où différents scopes fonctionnels qui aurait des modélisations locales)

Sur Wikipedia :

> DTO does not have any behavior except for storage, retrieval, serialization and deserialization of its own data (mutators, accessors, parsers and serializers). In other words, DTOs are simple objects that should not contain any business logic but may contain serialization and deserialization mechanisms for transferring data over the wire.

Reply:
Du coup, si je comprends bien le problème avec le terme de `DTO` c'est qu'il est surtout utilisé lors d'échanges client/serveur: ce qu'on reçoit du front ou envoie au front est un `DTO`. Par contre, en interne, ce qui traverse une couche serait plutôt un `Record`. Personnellement le terme `Record` ne me dérange pas si ça convient au plus grand nombre.